src/Facilitator/Facilitator.js

```diff
@@ -69,95 +79,637 @@ class Facilitator {
+   * @param {string} nonce Stake nonce.
+   * @param {string} hashLock Hash lock.
+   * @param {string} unlockSecret Unlock secret.
+   * @param {Object} txOptions Transaction options.
+   *
+   * @returns {Promise} Promise object.
+   */
+  confirmStakeIntent(
+    staker,
+    amount,
+    beneficiary,
+    gasPrice,
+    gasLimit,
+    nonce,
+    hashLock,
+    unlockSecret,
+    txOptions,
+  ) {
+    return new Promise(async (onResolve, onReject) => {
+      logger.info('Confirming stake intent');
+      logger.info('-----------------------');
+      if (!Web3.utils.isAddress(staker)) {
+        const err = new TypeError('Invalid staker address.');
+        onReject(err);
+      }
+      if (new BN(amount).eqn(0)) {
+        const err = new TypeError('Stake amount must not be zero.');
+        onReject(err);
+      }
+      if (!Web3.utils.isAddress(beneficiary)) {
+        const err = new TypeError('Invalid beneficiary address.');
+        onReject(err);
+      }
+      if (typeof gasPrice !== 'string') {
+        const err = new TypeError('Invalid gas price.');
+        onReject(err);
+      }
+      if (typeof gasLimit !== 'string') {
+        const err = new TypeError('Invalid gas limit.');
+        onReject(err);
+      }
+      if (typeof nonce !== 'string') {
+        const err = new TypeError('Invalid staker nonce.');
+        onReject(err);
+      }
+      if (!txOptions) {
+        const err = new TypeError('Invalid transaction options.');
+        onReject(err);
+      }
+      if (!Web3.utils.isAddress(txOptions.from)) {
+        const err = new TypeError('Invalid facilitator address.');
+        onReject(err);
+      }
+
+      logger.info('* Generating message hash with given stake parameters *');
+      const messageHash = Message.getStakeMessageHash(
+        amount,
+        beneficiary,
+        this.gateway.gatewayAddress,
+        nonce,
+        gasPrice,
+        gasLimit,
+        staker,
+        hashLock,
+      );
+
+      logger.info(`  - Message hash is ${messageHash}`);
+
+      const stakeMessageStatus = await this.gateway.getOutboxMessageStatus(
+        messageHash,
+      );
+
+      logger.info(
+        `  - Gateway's outbox message hash is ${stakeMessageStatus}`,
+      );
+
+      if (stakeMessageStatus === MessageStatus.UNDECLARED) {
+        const err = new TypeError('Stake message hash must be declared.');
+        onReject(err);
+      }
+
+      const mintMessageStatus = await this.coGateway.getInboxMessageStatus(
+        messageHash,
+      );
+
+      logger.info(
+        `  - CoGateway's inbox message hash is ${mintMessageStatus}`,
+      );
+
+      if (
+        mintMessageStatus === MessageStatus.DECLARED ||
+        mintMessageStatus === MessageStatus.PROGRESSED ||
+        mintMessageStatus === MessageStatus.REVOKED
+      ) {
+        logger.info('  - Stake intent already confirmed on CoGateway');
+        onResolve(true);
```

Comment:
If we return early here we don't need following logic to be nested in an else block.